### PR TITLE
HPCC-26826 Move flow control packet processing onto data read thread

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: roxie/udplib/udptrr.cpp [hunk 1]
```diff
@@ -415,20 +431,122 @@ class CReceiveManager : implements IReceiveManager, public CInterface
             flow_socket->set_receive_buffer_size(udpFlowSocketsSize);
             size32_t actualSize = flow_socket->get_receive_buffer_size();
             DBGLOG("UdpReceiver: receive_receive_flow created port=%d sockbuffsize=%d actual %d", flow_port, udpFlowSocketsSize, actualSize);
+            Thread::start();
         }
-        
-        ~receive_receive_flow() 
+
+        void doFlowRequest(const UdpRequestToSendMsg msg)
```
Comment (ghalliday) [inline roxie/udplib/udptrr.cpp:437]: reference parameter?

**Review by ghalliday — COMMENTED** [review]
@richardkchapman this also needs changes in ccdqueue.cpp if sendFlowOnDataPort.

I'm not sure why regression test would have passed unless localSlave was set.

**Comment by ghalliday** [comment]
trivial: could also remove this code from the catch handlers:
```
                    ::Release(b);
                    b = NULL;

```

**Review by mckellyln — APPROVED** [review]
Looks good to me.
Should I try a test on the large cluster ?

File: roxie/udplib/udptrr.cpp [hunk 2]
```diff
@@ -643,46 +647,85 @@ class CReceiveManager : implements IReceiveManager, public CInterface
...
-                    b = bufferManager->allocate();
-                    while (true)
+                    bool dataAvail = receive_socket->wait_read(timeout);
+                    if (dataAvail)
                     {
-                        receive_socket->read(b->data, 1, DATA_PAYLOAD, res, 5);
+                        if (!b)
+                            b = bufferManager->allocate();
+                        receive_socket->readtms(b->data, 1, DATA_PAYLOAD, res, 0);
                         if (res!=sizeof(UdpRequestToSendMsg))
```
Comment (mckellyln) [inline roxie/udplib/udptrr.cpp:666]: Is a data send always guaranteed to be larger than sizeof(udpRequestToSendMsg) ?

Comment (richardkchapman) [reply roxie/udplib/udptrr.cpp:666]: Yes - and I think I put in a static assert to confirm that.

File: roxie/udplib/udptrr.cpp (hunk 1 above)
Comment (richardkchapman) [reply roxie/udplib/udptrr.cpp:437]: Will change

**Comment by richardkchapman** [comment]
@ghalliday Please re-review

**Comment by ghalliday** [comment]
The conclusion from further discussions is that this is probably a bad idea because flow control messages from other servers are stuck behind the data packets.  Closing for the moment.

**Pull request closed without merging**